### PR TITLE
[AArch64] Implement __builtin_cpu_supports, compiler-rt tests.

### PR DESCRIPTION
**Pull request by ilinpv**
The patch complements https://github.com/llvm/llvm-project/pull/68919 and adds AArch64 support for builtin
`__builtin_cpu_supports("feature1+...+featureN")`
which return true if all specified CPU features in argument are detected. Also compiler-rt aarch64 native run tests for features detection mechanism were added and 'cpu_model' check was fixed after its refactor merged https://github.com/llvm/llvm-project/pull/75635 Original RFC was https://reviews.llvm.org/D153153

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -14025,6 +14028,19 @@ Value *CodeGenFunction::EmitX86CpuInit() {
   return Builder.CreateCall(Func);
 }
 
+Value *CodeGenFunction::EmitAArch64CpuSupports(const CallExpr *E) {
+  const Expr *ArgExpr = E->getArg(0)->IgnoreParenCasts();
+  StringRef ArgStr = cast<StringLiteral>(ArgExpr)->getString();
```
Comment (jroelofs) [inline clang/lib/CodeGen/CGBuiltin.cpp:14033]: Does prior machinery in sema for the builtin forbid e.g:

```
const char *feats = "rng+flagm";
__builtin_cpu_supports(feats);
```

If not, this cast may fail.

Comment (jroelofs) [reply clang/lib/CodeGen/CGBuiltin.cpp:14033]: oh, yes. it does.

**Review by jroelofs — APPROVED** [review]

**Comment by asmok-g** [comment]
Hello, 

This patch makes the following line fail to compile
```
#elif ABSL_HAVE_BUILTIN(__builtin_cpu_supports)
  if (__builtin_cpu_supports("avx2")) {
```

`error: invalid cpu feature string for builtin` when built on arm

**Comment by alexfh** [comment]
This makes the clang diverge from the GCC behavior, for example, GCC happily compiles `(void)__builtin_cpu_supports("neon");` on x86, but clang doesn't after this patch: https://gcc.godbolt.org/z/a367Pd7sE

**Comment by eaeltsin** [comment]
@ilinpv the problem is that a lot of third party code out there might fail to compile now because it assumes GCC's behavior for invalid `__builtin_cpu_supports`. At least we can see many breakages in the third party that we use in our codebase.

By intentionally diverging from GCC this way you assume third party code owners will have to add architecture-specific guards around each use of `__builtin_cpu_supports`, which might be pretty disruptive.

Is there any document that specifies how builtins should behave when they are not supported/their parameter is not supported? I assume there is a reason behind failing the compilation in these cases, as not every builtin can simply answer "no", but I think this should be documented very explicitly.

**Comment by DanielKristofKiss** [comment]
Backward\forward compatibility force us to support any parameter in `__builtin_cpu_supports `. 
A given codebase and toolchain add support for a new feature and everything works but we can't assume all cases the toolchain is up to date where the project is used so an older version of the compiler needs to accept the new feature string when it doesn't know about it. 
Also target and such a preprocessor checks will not help in this case.

**Comment by eaeltsin** [comment]
Thanks @DanielKristofKiss - considering compilations with older toolchains is a good reason to choose the particular behavior. It would be very valuable if all these reasons and the selected behavior for unsupported builtins/parameters are documented, i.e. here - https://clang.llvm.org/docs/LanguageExtensions.html#builtin-functions

**Comment by bgra8** [comment]
@DanielKristofKiss do I understand correctly that you confirm the current behavior for `__builtin_cpu_supports("unsupported-features")` -- does not compile no matter the CPU --  is a `clang` defect? 

If that is the case let's file a bug to have this fixed.

**Comment by DanielKristofKiss** [comment]
@bgra8 Correct, submitted #83407.

**Comment by ilinpv** [comment]
Fix on review https://github.com/llvm/llvm-project/pull/83515
The documentation is coming next.

**Comment by eaeltsin** [comment]
Thanks @ilinpv !

**Comment by ilinpv** [comment]
Documentation update https://github.com/llvm/llvm-project/pull/84098

**Pull request merged**